### PR TITLE
[OMNIML-3015]Add per tensor/per channel MSE calibrator

### PR DESCRIPTION
**Pull request by Fridah-nv**
## What does this PR do?

**Type of change:** ? 
new feature

**Overview:** ?

Add per tensor/per channel MSE calibrator.

## Usage
Can be enabled with "algorithm" field in quantization configs.
```
"algorithm": {"method": "mse", "num_steps": 20, "stop_multiplier": 8.0},
```

## Testing

Unit test for the MseCalibrator,
E2E test with NVFP4 and INT8,

**results: **
start_multiplier=0.25
stop_multiplier=4.0
num_steps=20

**Qwen3-8B MMLU:**

**BF16 baseline: 72.94**
| Calib Algo | NVFP4 | FP8 | INT8 |
| ------ | ------ | ------ | ------ |
|    MSE    |    70.88    |    72.65    |    55.46    |
|    MAX    |    70.83    |    72.7    |    24.52    |

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes
- **Did you write any new necessary tests?**: Yes
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No 

## Additional Information
TODO: for the follow up PR:
- [ ] TP sync for HF models
- [ ] Calculate weight quantizer only once

## Release Notes

* **New Features**
  * Added MSE-based quantization calibration supporting per-tensor and per-channel optimization with configurable parameters (step count, multiplier ranges).

* **Tests**
  * Added comprehensive test coverage for MSE calibration functionality.

* **Documentation**
  * Updated changelog to reflect MSE calibrator support.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: modelopt/torch/quantization/calib/mse.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+            if reduce_axis is None:
+                loss = torch.sum(error)
+            else:
+                loss = quant_utils.reduce_sum(error, axis=reduce_axis, keepdims=False)
+
+            if self._candidate_amaxs[step] is None:
+                self._candidate_amaxs[step] = candidate_amax
+
+            self._losses[step].append(loss)
```
Comment (realAsma) [inline modelopt/torch/quantization/calib/mse.py]: can we do 
```
self._losses[step] += loss
```
self._losses[step].append(loss) could consume too much memory

**Comment by realAsma** [comment]
@coderabbitai review

File: modelopt/torch/quantization/calib/mse.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            loss = quant_utils.reduce_sum(error, axis=reduce_axis, keepdims=False)
+
+            if self._candidate_amaxs[step] is None:
+                self._candidate_amaxs[step] = candidate_amax
+
+            if self._losses_sum[step] is None:
+                self._losses_sum[step] = loss.clone()
+            else:
+                self._losses_sum[step] += loss
+            self._losses_count[step] += 1
```
Comment (realAsma) [inline modelopt/torch/quantization/calib/mse.py]: why do we need a super why do we need this self._losses_count variable? looks like all steps will have same  self._losses_count

File: modelopt/torch/quantization/calib/mse.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+                break
+
+        if first_loss_sum is None:
+            return None
+
+        if first_loss_sum.ndim == 0:
+            avg_losses = []
+            for step in range(self._num_steps):
+                if self._losses_sum[step] is not None and self._losses_count[step] > 0:
+                    avg_loss = self._losses_sum[step] / self._losses_count[step]
```
Comment (realAsma) [inline modelopt/torch/quantization/calib/mse.py]: averaging of loss does not seem to be needed. Why dont simply take search on self._losses_sum[step]?

File: modelopt/torch/quantization/calib/mse.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        first_loss_sum = None
+        for loss_sum in self._losses_sum:
+            if loss_sum is not None:
+                first_loss_sum = loss_sum
+                break
+
+        if first_loss_sum is None:
+            return None
+
+        if first_loss_sum.ndim == 0:
```
Comment (realAsma) [inline modelopt/torch/quantization/calib/mse.py]: nit:
can we merge the per-tensor case search and per-channel case search ? There seems to be duplication of code.

File: modelopt/torch/quantization/model_calib.py [hunk 5]
```diff
@@ -181,6 +185,88 @@ def sync_quantizer_amax_across_tp(
...
+
+                    with (
+                        enable_quant(quantizer),
+                        disable_calib(quantizer),
+                        enable_fake_quant(quantizer),
+                    ):
+                        xq = quantizer(x)
+
+                    if original_amax is not None:
+                        quantizer._amax = original_amax
```
Comment (realAsma) [inline modelopt/torch/quantization/model_calib.py:243]:
```suggestion
                    if original_amax is not None:
                        quantizer._amax = original_amax
                    else:
                        delattr(quantizer, "_amax")
```

File: modelopt/torch/quantization/nn/modules/tensor_quantizer.py [hunk 6]
```diff
@@ -626,7 +635,7 @@ def _fake_quantize(self, inputs):
             amax = self._get_amax(inputs)
 
         if self.block_sizes is not None and self.block_sizes.get("type", "static") == "dynamic":
-            # Block quantization, including dynamic and static block quantization
+            # Dynamic block quantization
```
Comment (realAsma) [inline modelopt/torch/quantization/nn/modules/tensor_quantizer.py:638]: thanks for the fix,

File: modelopt/torch/quantization/calib/mse.py (hunk 3 above)
Comment (Fridah-nv) [reply modelopt/torch/quantization/calib/mse.py]: Thanks for the suggestion! I Also deleted `self._losses_count`

**Review by realAsma — APPROVED** [review]
Looks great!!

Can we also have an end to end unittest of the flow in `tests/unit/torch/quantization/test_quantize_cpu.py::test_quantize`?
